### PR TITLE
#184557534 Implement displaying sales stats

### PR DESCRIPTION
**Pull request by Munezeroolivierhugue**
### What does this PR do?

Displays statistics about sales

### Description of Task to be completed?

- Getting sales stats
- Getting expired product stats
- Getting loss on expired product
- Getting balance after loss and sales

### How should this be manually tested?

1. Clone [repo](https://github.com/atlp-rwanda/vikings-ec-bn.git)
2. Change into the directory vikings-ec-bn
3. Checkout to the branch: feature/sales-stats-184557534
4. Run the command `npm install` to install all packages
5. Run `npm run seeding`
6. Run `npm run dev`
7. Get product stats at `http://localhost:<port>/api/v1/stats/sales-stats?<query>`  URL in Postman

#### What are the relevant pivotal tracker stories?

[#184557534](https://www.pivotaltracker.com/story/show/184557534)

### CODE REVIEW

**Review by fabmukunzi — APPROVED** [review]

**Review by irakozetony — APPROVED** [review]

**Review by ndatumuremyi — APPROVED** [review]

File: src/controllers/stats.controller.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    }
+  }
+}
+
+export const knownStats = {
+  wishes:'wishes',
+  sales: 'sales',
+  expiredProduct:'expired',
+  productCreated: 'product-created',
+};
```
Comment (ncutixavier) [inline src/controllers/stats.controller.js]: This is not controller

File: src/middlewares/auth.middleware.js [hunk 2]
```diff
@@ -38,6 +38,8 @@ export const protectRoute = async (req, res, next) => {
 };
 export const restrictTo = (...roles) => {
   return (req, res, next) => {
+    console.log(roles,"roles")
+    console.log(req.user.role,"req roles")
```
Comment (ncutixavier) [inline src/middlewares/auth.middleware.js]: Remove all unnecessary logs

File: src/middlewares/stats.middleware.js [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    next();
+}
+export const getSales = async (req, res, next) => {
+    if(req.statsFields.includes(knownStats.sales)){
+        const start = req.query.start;
+      const end = req.query.end;
+        const sellerId = req.user.id;
+
+        const stats = await SalesService.getAllSalesStats(start, end);
+      // console.log(stats, "stats");
```
Comment (ncutixavier) [inline src/middlewares/stats.middleware.js]: Remove all unnecessary comments

File: src/middlewares/stats.middleware.js [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+              .catch(reject);
+          });
+        })
+      );
+
+      const productS = productWithSales.filter(each => {
+        return each.product.userId === sellerId
+      }).map(each => {
+        return each;
+      }).reduce((pv,cv) => {
```
Comment (ncutixavier) [inline src/middlewares/stats.middleware.js]: What is pv, cv, productS

File: src/routes/api/stats.routes.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import express from "express";
+import { StatsController } from "../../controllers/stats.controller";
+import { protectRoute, restrictTo } from "../../middlewares/auth.middleware";
+import { statsValidation } from "../../validations/stats/stats.validation";
+import {validateStats, getSales, getWishes, getExpired, getProductCreated} from "../../middlewares/stats.middleware";
+
+const statsRouter = express.Router();
+
+statsRouter.get(
+  "/sales-stats",
```
Comment (ncutixavier) [inline src/routes/api/stats.routes.js]: Update this route

**Review by ncutixavier — CHANGES_REQUESTED** [review]

File: src/middlewares/stats.middleware.js [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+            }catch(e){
+              reject(e);
+            }
+          })();
+          
+        });
+      }));
+      let productWithWishes = sellerWishes.reduce((prev, currentWish) => {
+        const productWished = currentWish.products.forEach(each => {
+          const eachW = {id: each.id, name: each.name}
```
Comment (irakozetony) [inline src/middlewares/stats.middleware.js]: Rename the variables to make them more clear and understandable (eachW)

File: src/middlewares/stats.middleware.js [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+      }, {});
+      const productStatistics = Object.keys(productSalesStats).map(each =>{
+        return {
+          productId: each,
+          ...productSalesStats[each]
+        }
+      })
+         
+      const salesbycategory = productStatistics.reduce((prev,curr) => {
+        if(!prev[curr.categoryId]){
```
Comment (irakozetony) [inline src/middlewares/stats.middleware.js]: Rename variables here too! (prev & curr)

**Review by irakozetony — CHANGES_REQUESTED** [review]

**Review by Abdden — APPROVED** [review]

**Review by irakozetony — APPROVED** [review]

**Review by ncutixavier — APPROVED** [review]

**Comment by ncutixavier** [comment]
Update your commit message

**Pull request merged**